### PR TITLE
adding session properties 

### LINKED ISSUE #1203: Add session properties for per-node memory limits
New properties:

* `query_max_memory_per_node` (config `query.max-memory-per-node`)
* `query_max_total_memory_per_node (config `query_max_total_memory`)

These are similar to the existing session properties for global query limits:

* `query_max_memory`
* `query_max_total_memory`

### PR DESCRIPTION
**Pull request by ankitdixit**
for queryMaxMemoryPerNode and queryMaxTotalMemoryPerNode

Fixes #1203

### CODE REVIEW

**Comment by Terry93** [comment]
thank you for the PR!

File: presto-tests/src/test/java/io/prestosql/memory/TestMemoryManager.java [hunk 1]
```diff
@@ -327,6 +329,50 @@ public void testClusterPools()
         }
     }
 
+    @Test(timeOut = 240_000)
+    public void testSessionQueryMemoryPerNodeLimit()
+            throws Exception
+    {
+        try (DistributedQueryRunner queryRunner = createQueryRunner(TINY_SESSION, ImmutableMap.of())) {
```
Comment (sopel39) [inline presto-tests/src/test/java/io/prestosql/memory/TestMemoryManager.java]: You should run two queries. One that passes without memory limit and one that fails once you apply memory limit session property. Take a look at `testResourceOverCommit` test.

File: presto-tests/src/test/java/io/prestosql/memory/TestMemoryManager.java [hunk 2]
```diff
@@ -327,6 +329,50 @@ public void testClusterPools()
...
+                // expected
+            }
+        }
+    }
+
+    @Test(timeOut = 240_000)
+    public void testSessionQueryMaxTotalMemoryPerNodeLimit()
+            throws Exception
+    {
+        try (DistributedQueryRunner queryRunner = createQueryRunner(TINY_SESSION, ImmutableMap.of())) {
```
Comment (sopel39) [inline presto-tests/src/test/java/io/prestosql/memory/TestMemoryManager.java]: ditto

File: presto-main/src/main/java/io/prestosql/SystemSessionProperties.java [hunk 3]
```diff
@@ -19,6 +19,7 @@
 import io.prestosql.execution.QueryManagerConfig;
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/SystemSessionProperties.java]: Commit message is too long.

File: presto-tests/src/test/java/io/prestosql/memory/TestMemoryManager.java (hunk 1 above)
Comment (sopel39) [inline presto-tests/src/test/java/io/prestosql/memory/TestMemoryManager.java]: please share query runner across different tests. Query runner takes time to create

Comment (Praveen2112) [reply presto-tests/src/test/java/io/prestosql/memory/TestMemoryManager.java]: @sopel39 How about extend `AbstractTestQueryFramework` for these tests ?

Comment (ankitdixit) [reply presto-tests/src/test/java/io/prestosql/memory/TestMemoryManager.java]: @sopel39 I do run the query twice at line 337 and line 345. Or I might be misunderstanding the ask.

Comment (ankitdixit) [reply presto-tests/src/test/java/io/prestosql/memory/TestMemoryManager.java]: I tried extending AbstractTestQueryFramework but it seems to be a generic framework with a lot of tests which are not applicable here.
Furthermore, all the tests set properties while creating query runner so they require creation of different query runners.

Comment (ankitdixit) [reply presto-tests/src/test/java/io/prestosql/memory/TestMemoryManager.java]: discussed offline, this is already written in the suggested format

File: presto-tests/src/test/java/io/prestosql/memory/TestMemoryManager.java (hunk 2 above)
Comment (ankitdixit) [reply presto-tests/src/test/java/io/prestosql/memory/TestMemoryManager.java]: discussed offline

File: presto-main/src/main/java/io/prestosql/SystemSessionProperties.java (hunk 3 above)
Comment (ankitdixit) [reply presto-main/src/main/java/io/prestosql/SystemSessionProperties.java]: changed

File: presto-tests/src/test/java/io/prestosql/memory/TestMemoryManager.java (hunk 1 above)
Comment (sopel39) [reply presto-tests/src/test/java/io/prestosql/memory/TestMemoryManager.java]: `AbstractTestQueryFramework` doesn't have any tests. You can use methods that accept `Session` as it's parameter, e.g. `io.prestosql.tests.AbstractTestQueryFramework#assertQuery(io.prestosql.Session, java.lang.String, java.lang.String)`

File: presto-main/src/main/java/io/prestosql/execution/SqlTaskManager.java [hunk 4]
```diff
@@ -353,6 +361,17 @@ public TaskInfo updateTask(Session session, TaskId taskId, Optional<PlanFragment
         requireNonNull(sources, "sources is null");
         requireNonNull(outputBuffers, "outputBuffers is null");
 
+        long sessionQueryMaxMemoryPerNode = getQueryMaxMemoryPerNode(session).toBytes();
+        long sessionQueryTotalMaxMemoryPerNode = getQueryMaxTotalMemoryPerNode(session).toBytes();
+        // Session property query_max_memory_per_node is used to only decrease memory limit
+        if (sessionQueryMaxMemoryPerNode < queryMaxMemoryPerNode) {
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/execution/SqlTaskManager.java]: I think it should be `<=`

File: presto-main/src/main/java/io/prestosql/execution/SqlTaskManager.java [hunk 5]
```diff
@@ -353,6 +361,17 @@ public TaskInfo updateTask(Session session, TaskId taskId, Optional<PlanFragment
...
+            queryContexts.getUnchecked(taskId.getQueryId()).setMaxUserMemory(sessionQueryMaxMemoryPerNode);
+        }
+
+        if (sessionQueryTotalMaxMemoryPerNode < queryMaxTotalMemoryPerNode) {
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/execution/SqlTaskManager.java]: ditto

File: presto-tests/src/main/java/io/prestosql/tests/DistributedQueryRunner.java [hunk 6]
```diff
@@ -467,7 +467,7 @@ private static void closeUnchecked(AutoCloseable closeable)
         private String environment = ENVIRONMENT;
         private Optional<Path> baseDataDir = Optional.empty();
 
-        protected Builder(Session defaultSession)
+        public Builder(Session defaultSession)
```
Comment (sopel39) [inline presto-tests/src/main/java/io/prestosql/tests/DistributedQueryRunner.java]: you can use `io.prestosql.tests.DistributedQueryRunner.Builder#amendSession` instead

File: presto-tests/src/test/java/io/prestosql/memory/TestMemorySessionProperties.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+    TestMemorySessionProperties()
+    {
+        super(() -> new DistributedQueryRunner.Builder(TINY_SESSION)
+                .setNodeCount(2)
+                .build());
+    }
+
+    @BeforeClass
+    public void init()
```
Comment (sopel39) [inline presto-tests/src/test/java/io/prestosql/memory/TestMemorySessionProperties.java]: You can use `io.prestosql.tests.tpch.TpchQueryRunnerBuilder` instead of explicitly creating catalogs

File: presto-tests/src/test/java/io/prestosql/memory/TestMemorySessionProperties.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+            // expected
+        }
+    }
+
+    @Test(timeOut = 240_000)
+    public void testSessionQueryMaxTotalMemoryPerNodeLimit()
+            throws Exception
+    {
+        assertQuery(sql);
+        Session session = testSessionBuilder()
```
Comment (sopel39) [inline presto-tests/src/test/java/io/prestosql/memory/TestMemorySessionProperties.java]: You can use `Session.builder(getQueryRunner().getDefaultSession())` instead, so you don't need to set catalog and schema

File: presto-tests/src/test/java/io/prestosql/memory/TestMemorySessionProperties.java (hunk 7 above)
Comment (ankitdixit) [reply presto-tests/src/test/java/io/prestosql/memory/TestMemorySessionProperties.java]: yes, I was in 2 minds about using that since it was not used in another class where it could have been. Making this change

File: presto-tests/src/test/java/io/prestosql/memory/TestMemorySessionProperties.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+public class TestMemorySessionProperties
+        extends AbstractTestQueryFramework
+{
+    public static final String sql = "SELECT COUNT(*), clerk FROM orders GROUP BY clerk";
+
+    TestMemorySessionProperties()
+    {
+        super(() -> TpchQueryRunnerBuilder.builder()
+                        .amendSession(builder -> builder.setCatalog("tpch").setSchema("tiny"))
```
Comment (sopel39) [inline presto-tests/src/test/java/io/prestosql/memory/TestMemorySessionProperties.java]: remove, that is the default

File: presto-tests/src/test/java/io/prestosql/memory/TestMemoryManager.java [hunk 10]
```diff
@@ -379,19 +393,13 @@ public void testQueryMemoryPerNodeLimit()
         }
     }
 
-    public static DistributedQueryRunner createQueryRunner(Session session, Map<String, String> properties)
+    public static DistributedQueryRunner createQueryRunner(Session session, Map<String, String> extraProperties)
             throws Exception
     {
-        DistributedQueryRunner queryRunner = new DistributedQueryRunner(session, 2, properties);
```
Comment (sopel39) [inline presto-tests/src/test/java/io/prestosql/memory/TestMemoryManager.java]: nit: you could move it to separate commit

File: presto-tests/src/test/java/io/prestosql/memory/TestMemoryManager.java [hunk 11]
```diff
@@ -350,6 +351,19 @@ public void testQueryUserMemoryLimit()
                 .build();
         try (QueryRunner queryRunner = createQueryRunner(SESSION, properties)) {
             queryRunner.execute(SESSION, "SELECT COUNT(*), repeat(orderstatus, 1000) FROM orders GROUP BY 2");
+
+            Session session = testSessionBuilder()
```
Comment (sopel39) [inline presto-tests/src/test/java/io/prestosql/memory/TestMemoryManager.java]: this seems redundant to `testSessionQueryMemoryPerNodeLimit` test, isn't it?

Comment (ankitdixit) [reply presto-tests/src/test/java/io/prestosql/memory/TestMemoryManager.java]: yes, this is redundant, removing it.

File: presto-tests/src/test/java/io/prestosql/memory/TestMemoryManager.java (hunk 10 above)
Comment (ankitdixit) [reply presto-tests/src/test/java/io/prestosql/memory/TestMemoryManager.java]: moved

**Comment by kekwan** [comment]
@ankitdixit @sopel39 Why was this implemented so it only allows decreasing? Are there any technical reasons or barriers in that it's not safe to have it be increasing?

We have a use-case where we want allow select users overriding our cluster's global `query.max-memory-per-node` limit. We are looking in making the change into our fork but want to understand it better first.
